src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/MultiphasePoromechanics_impl.hpp

```diff
@@ -214,48 +217,26 @@ computeBodyForce( localIndex const k,
                   StackVariables & stack,
                   FUNC && bodyForceKernelOp ) const
 {
-  using Deriv = constitutive::multifluid::DerivativeOffset;
-
   GEOS_UNUSED_VAR( dPorosity_dTemperature );
 
-  arraySlice1d< real64 const, constitutive::multifluid::USD_PHASE - 2 > const phaseMassDensity = m_fluidPhaseMassDensity[k][q];
-  arraySlice2d< real64 const, constitutive::multifluid::USD_PHASE_DC - 2 > const dPhaseMassDensity = m_dFluidPhaseMassDensity[k][q];
-  arraySlice1d< real64 const, compflow::USD_PHASE - 1 > const phaseVolFrac = m_fluidPhaseVolFrac[k];
-  arraySlice2d< real64 const, compflow::USD_PHASE_DC - 1 > const dPhaseVolFrac = m_dFluidPhaseVolFrac[k];
-  arraySlice2d< real64 const, compflow::USD_COMP_DC - 1 > const dGlobalCompFrac_dGlobalCompDensity = m_dGlobalCompFraction_dGlobalCompDensity[k];
-
   // Step 1: compute fluid total mass density and its derivatives
 
   real64 totalMassDensity = 0.0;
-  real64 dTotalMassDensity_dPressure = 0.0;
   real64 dTotalMassDensity_dComponents[maxNumComponents]{};
-  real64 dPhaseMassDensity_dComponents[maxNumComponents]{};
 
-  for( integer ip = 0; ip < m_numPhases; ++ip )
+  arraySlice1d< real64 const, compflow::USD_COMP - 1 > const compDens = m_compDens[k];
+  for( integer ic = 0; ic < m_numComponents; ++ic )
   {
-    totalMassDensity += phaseVolFrac( ip ) * phaseMassDensity( ip );
-    dTotalMassDensity_dPressure += dPhaseVolFrac( ip, Deriv::dP ) * phaseMassDensity( ip )
-                                   + phaseVolFrac( ip ) * dPhaseMassDensity( ip, Deriv::dP );
-
-    applyChainRule( m_numComponents,
-                    dGlobalCompFrac_dGlobalCompDensity,
-                    dPhaseMassDensity[ip],
-                    dPhaseMassDensity_dComponents,
-                    Deriv::dC );
-    for( integer jc = 0; jc < m_numComponents; ++jc )
-    {
-      dTotalMassDensity_dComponents[jc] += dPhaseVolFrac( ip, Deriv::dC+jc ) * phaseMassDensity( ip )
-                                           + phaseVolFrac( ip ) * dPhaseMassDensity_dComponents[jc];
-    }
+    totalMassDensity += m_useMass ? compDens[ic] : compDens[ic] * m_componentMolarWeight[ic];
```

Comment:
we should avoid `+=` in kernels.

Reply:
Based on the offline discussion we had and on the test on Godbolt (thank you @klevzoff) this is not needed.

Reply:
The kernel optimization efforts for matrix-free on mechanics showed that using the `+=` did not result in an FMA across the `+=` operation. This was with `cuda10.1.243`. This was confirmed by the Nvidia staff at the hackathon who conferred with some others at vida. Maybe Nvidia fixed it...maybe it was specific to the architecture. I cannot tell you why it works for the Godbolt example, and it didn't work back then.

Reply:
Done